lib/instrumentation/grpc-js/grpc.js

```diff
@@ -6,80 +6,206 @@
+
+    const name = args[0]
+    const handler = args[1]
+    const type = args[args.length - 1]
+    // We need the server for the host and port, but depending on how
+    // the user calls server functions, the connexion might not be
+    // ready yet. It will be ready below by the time the registered
+    // handler is invoked.
+    const server = this
+
+    if (this.handlers.has(name)) {
+      shim.logger.debug(`gRPC handler for ${name} won't be added so not instrumenting`)
+      return original.apply(this, arguments)
+    }
 
-          originalListener && originalListener.onReceiveStatus(status)
-        }
+    args[1] = shim.bindCreateTransaction(createTransaction, { type: shim.WEB })
 
-        args[1] = nrListener
+    return original.apply(this, args)
 
-        return original.apply(this, args)
-      }
+    function createTransaction(stream) {
+      const parent = shim.getActiveSegment()
+      const transaction = parent.transaction
+      const address = server.http2ServerList[0].channelzRef.name
+
+      // Create the transaction segment using the request URL for now. Once a
+      // better name can be determined this segment will be renamed to that.
+      const segment = shim.createSegment(name, recordHttp)
+      segment.start()
+
+      transaction.type = 'web'
+      transaction.baseSegment = segment
+      transaction.url = name
+
+      transaction.trace.attributes.addAttribute(DESTINATION, 'request.method', transaction.url)
+
+      transaction.trace.attributes.addAttribute(
+        DESTINATION,
+        'request.uri',
+        `grpc://${address}${transaction.url}`
+      )
+
+      shim.setTransactionUri(transaction.url)
+
+      // This attribute isn't required by the spec, but it seems useful to have
+      transaction.trace.attributes.addAttribute(DESTINATION, 'grpc.type', type)
+
+      const metadata = stream.metadata
+      Object.entries(metadata.getMap()).forEach(([key, value]) => {
+        transaction.trace.attributes.addAttribute(DESTINATION, `request.headers.${key}`, value)
+      })
+
+      const headers = Object.create(null)
+      headers.tracestate = metadata.get('tracestate').join(',')
+      headers.traceparent = metadata.get('traceparent').join(',')
+      headers.newrelic = metadata.get('newrelic').join(',')
+      transaction.acceptDistributedTraceHeaders('HTTP', headers)
+
+      const agent = shim.agent
+      stream.call.once('callEnd', (statusCode) => {
+        transaction.trace.attributes.addAttribute(DESTINATION, 'response.status', statusCode)
+        if (statusCode !== 0 && agent.config.grpc.record_errors) {
+          const status = constants.Status[statusCode]
+          const error = new Error(`gRPC status code ${statusCode}: ${status}`)
+          agent.errors.add(transaction, error)
+        }
+      })
+      stream.call.once('streamEnd', () => {
+        transaction.end()
+      })
+      // TODO should also instrument the 'data' event on the stream
```

Comment:
do we still need to do this?

Reply:
I think it's worthwhile for a future implementation. If we have a customer who does a lot of gRPC streaming, they may be surprised if their client streaming isn't showing up in traces.

Reply:
ok, it's prob best to log a ticket to capture this

Reply:
Okay, done.